New feature that notifies a user when their group invite is accepted.

Hus274 opened the pull request:
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #7788 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
Created a new feature that notifies the inviting user when another use accepts their invitation to join a group. This works for both parties and guilds. I've also added tests to support this new behavior.

<img width="269" alt="screen shot 2016-11-29 at 3 20 30 pm" src="https://cloud.githubusercontent.com/assets/17734753/20733321/9b286fa0-b647-11e6-9c1f-f008844588de.png">

This is in replacement of the previous pull request #7797 which I closed due to the website refactoring.

Hus274 committed:
New notification feature that notifies a user when their group invite is accepted. fixes #7788

crookedneighbor commented:
This might have been discussed in the previous PR, but why use the easy to miss blue notification instead of the modal notifications stored in user.notifications?

Hus274 commented:
That's because it seems to be what paglias was suggesting in the last one:
> I was thinking more about a one time blue message in the top right that you can trigger from the Notification service.

I get what you're saying about it being easy to miss and may make more sense as a modal and I'm alright with that direction as well.

Hus274 committed:
Updating to a modal instead of a popup notification

Hus274 commented:
I've updated this to use a modal instead:

<img width="609" alt="screen shot 2016-11-29 at 5 57 24 pm" src="https://cloud.githubusercontent.com/assets/17734753/20736947/a07bdde0-b65e-11e6-8bfe-70f1535c3ddb.png">

I figured the diff is pretty minimal so i can just revert if we don't want to use a modal.

crookedneighbor commented on website/client-old/js/controllers/notificationCtrl.js:
```diff
@@ -144,6 +144,11 @@ habitrpg.controller('NotificationCtrl',
           case 'LOGIN_INCENTIVE':
             Notification.showLoginIncentive(User.user, notification.data, Social.loadWidgets);
             break;
+          case 'GROUP_INVITE_ACCEPTED':
```
Do we not have a generic notification modal we use when the type isn't found on the client?

paglias replied:
I don't think so, we could add it though

crookedneighbor replied:
I think we should, rather than continually adding new modals.

paglias commented:
Yeah, it'll also make it easier to handle these notifications in other
clients like the mobile apps since the message will be provided by the
server instead of being specified in each modal source code

Il 30 Nov 2016 15:38, "Blade Barringer" <notifications@github.com> ha
scritto:

> *@crookedneighbor* commented on this pull request.
> ------------------------------
>
> In website/client-old/js/controllers/notificationCtrl.js
> <https://github.com/HabitRPG/habitica/pull/8244>:
>
> > @@ -144,6 +144,11 @@ habitrpg.controller('NotificationCtrl',
>            case 'LOGIN_INCENTIVE':
>              Notification.showLoginIncentive(User.user, notification.data, Social.loadWidgets);
>              break;
> +          case 'GROUP_INVITE_ACCEPTED':
>
> I think we should, rather than continually adding new modals.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/HabitRPG/habitica/pull/8244>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkAV4QPOAZ2B394KqjgfBvhPaiprlwdks5rDYpNgaJpZM4K_nK0>
> .
>

Hus274 commented:
I'm not sure I totally understand how to design what you're looking for here. Would the server just return the header and message for the modal and create a default case statement that takes any undefined notification type and assume it has a header and body and then just pass that into a generic modal that displays those fields?

paglias commented:
@Hus274 I think for generic notifications notification.data should have `{title, text}` and when a notification type hasn't a special handler a modal is shown with `data.title` in the header and `data.header` in the body

Hus274 commented:
@paglias Ok cool, that's about what I was understanding. I can take a look into making this generic then. Thanks!

Hus274 committed:
Making a generic modal template and using it for notifications of group invitation acceptance.

Hus274 commented:
I added a new generic modal template and made the invite accept notification use that instead. It allows the server to pass header and body text keys in addition to header and body data fields that can be used to fill in the translatable text to make it a little more flexible and reusable.

Let me know what you think.

Hus274 committed:
Working with paglias's comments for doing translation server side.

Hus274 commented:
Updated this for @paglias' comments. I've moved the translation to the server side and required that the generic modal popup only need the data object passed to it from the notification to have the translated headerText and bodyText in order for the modal to properly work.

Thanks for all the feedback @paglias, it was super helpful! Let me know if there's anything else that needs to get updated here!

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -303,6 +305,18 @@ api.joinGroup = {
+      let data = {
+        headerText: common.i18n.t('invitationAcceptedHeader', inviter.preferences.language),
+        bodyText: common.i18n.t('invitationAcceptedBody', {groupName: group.name, username: user.auth.local.username},
+          inviter.preferences.language),
```
I would put this back on the upper line and instead split the `{groupName: group.name, username: user.auth.local.username}` on multiple lines, like

```
data = {
  headerText: common.i18n.t('invitationAcceptedHeader', inviter.preferences.language),
  bodyText: common.i18n.t('invitationAcceptedBody', {
    groupName: group.name, 
    username: user.auth.local.username
  }, inviter.preferences.language),
}
```

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -303,6 +305,18 @@ api.joinGroup = {
+    }
+
     if (group.type === 'party' && inviter) {
       promises.push(User.update({_id: inviter}, {$inc: {'items.quests.basilist': 1}}).exec()); // Reward inviter
```
since we load the inviter, it's better to do the modification on the inviter object and then `save` instead of a separate update (you'll need to load `items.quests.basilist` other than `notifications preferences.language `

paglias commented on test/api/v3/integration/groups/POST-groups_groupId_join.test.js:
```diff
@@ -134,6 +134,14 @@ describe('POST /group/:groupId/join', () => {
+      it('notifies inviting user that their invitation was accepted', async () => {
+        await invitedUser.post(`/groups/${guild._id}/join`);
+
+        let inviter = await user.get('/user');
```
can we add expectation for the notification data too?

paglias commented:
@Hus274 I left 3 comments, but they're minor, otherwise it's very good and I'll merge it as soon as these last things are addressed. Thanks!

Hus274 committed:
Final changes based on pr comments.

Hus274 commented:
Made those final changes. Definitely a good call on expecting on the data output from the join method. Let me know if there's anything else that needs to be updated!

paglias commented:
Thanks! Notated towards your next tier
